components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/dao/impl/AuthenticatorManagementDAOImpl.java

```diff
@@ -0,0 +1,303 @@
+                    statement.setInt(Column.TENANT_ID, tenantId);
+                    statement.setString(Column.DEFINED_BY, DefinedByType.USER.toString());
+                });
+
+        if (authConfig == null) {
+            return null;
+        }
+
+        int authenticatorConfigID = getAuthenticatorIdentifier(authenticatorConfigName, tenantId);
+        List<Property> properties = new ArrayList<>();
+        jdbcTemplate.fetchSingleRecord(Query.GET_AUTHENTICATOR_PROP_SQL,
+                (resultSet, rowNumber) -> {
+                    Property property = new Property();
+                    property.setName(resultSet.getString(Column.PROPERTY_KEY));
+                    property.setValue(resultSet.getString(Column.PROPERTY_VALUE));
+                    property.setConfidential(false);
+                    properties.add(property);
+                    return null;
+                },
+                statementProp -> {
+                    statementProp.setInt(Column.AUTHENTICATOR_ID, authenticatorConfigID);
+                    statementProp.setInt(Column.TENANT_ID, tenantId);
+                });
+        authConfig.setProperties(properties.toArray(new Property[0]));
+        return authConfig;
+    }
+
+    private UserDefinedLocalAuthenticatorConfig getLocalAuthenticatorConfigBasedOnType(String authenticationType) {
+
+        if (AuthenticationType.VERIFICATION.toString().equals(authenticationType)) {
+            return new UserDefinedLocalAuthenticatorConfig(AuthenticationType.VERIFICATION);
+        }
+        return new UserDefinedLocalAuthenticatorConfig(AuthenticationType.IDENTIFICATION);
+    }
+
+    private int getAuthenticatorIdentifier(String authenticatorConfigName, int tenantId)
+            throws AuthenticatorMgtServerException, DataAccessException {
+
+        NamedJdbcTemplate jdbcTemplate = new NamedJdbcTemplate(IdentityDatabaseUtil.getDataSource());
+        String id = jdbcTemplate.fetchSingleRecord(Query.GET_AUTHENTICATOR_ID_SQL,
+            (resultSet, rowNumber) -> resultSet.getString(Column.ID),
+            statement -> {
+                statement.setString(Column.NAME, authenticatorConfigName);
+                statement.setInt(Column.TENANT_ID, tenantId);
+            });
+
+        if (id != null) {
+            return Integer.parseInt(id);
+        }
+        throw buildServerException(AuthenticatorMgtError.ERROR_CODE_NO_AUTHENTICATOR_FOUND,
+                    authenticatorConfigName);
+    }
+
+    private void addAuthenticatorProperties(String authenticatorName, int authenticatorConfigID, Property[] properties,
+                                            int tenantId) throws DataAccessException, AuthenticatorMgtServerException {
+
+        if (!(properties.length == 1 && ACTION_ID_PROPERTY.equals(properties[0].getName()))) {
+            throw buildServerException(AuthenticatorMgtError.ERROR_CODE_HAVING_MULTIPLE_PROP,
+                    authenticatorName);
+        }
```

Comment:
Better explain the validation or extract to a method to improve readability

Reply:
Ideally this validation should move to the layer above as this is related to business logic. Based on the way this is placed when this fails the authenticator seems to have been stored partially